### PR TITLE
[v12] Replace Webpack with Vite

### PR DESCRIPTION
**Pull request by ryanclark**
Backport Vite to v12

Original PR - https://github.com/gravitational/teleport/pull/21168
Includes changes from
- https://github.com/gravitational/teleport/pull/23427
- https://github.com/gravitational/teleport/pull/23815
- https://github.com/gravitational/teleport/pull/24591
- https://github.com/gravitational/teleport/pull/24904
- https://github.com/gravitational/teleport/pull/26084
- https://github.com/gravitational/teleport/pull/25514

### CODE REVIEW

**Review by kimlisa — APPROVED** [review]

File: package.json [hunk 1]
```diff
@@ -18,13 +18,14 @@
...
     "eslint": "eslint --quiet '+(e|web)/**/*.{ts,tsx,js,jsx}'",
-    "type-check": "tsc --noEmit",
+    "type-check": "NODE_OPTIONS='--max-old-space-size=4096' tsc --noEmit",
     "prettier-check": "yarn prettier --check '+(e|web)/**/*.{ts,tsx,js,jsx}'",
     "prettier-write": "yarn prettier --write --loglevel silent '+(e|web)/**/*.{ts,tsx,js,jsx}'",
     "nop": "exit 0"
   },
   "private": true,
   "resolutions": {
+    "@swc/core": "1.3.36",
```
Comment (ravicious) [inline package.json:28]: Why is this needed on v12 and not on master?

File: tsconfig.node.json [hunk 2]
```diff
@@ -0,0 +1,12 @@
+{
+  "compilerOptions": {
+    "composite": true,
+    "module": "ESNext",
+    "moduleResolution": "Node",
+    "allowSyntheticDefaultImports": true
+  },
+  "include": [
+    "web/packages/build/vite/*.ts",
+    "web/packages/teleport/vite.config.ts"
```
Comment (ravicious) [inline tsconfig.node.json:10]: I didn't notice that before, how does it decide which `tsconfig` to pick?

File: tsconfig.json
Comment (ravicious) [inline tsconfig.json:1]: It looks like the changes here cause some type errors to pop up in `web/packages/teleterm`. I wonder why it's the case on v12 and not on v13/master. 🤔

File: package.json (hunk 1 above)
Comment (ryanclark) [reply package.json:28]: On master/v13, the dependencies for swc are set in `yarn.lock` to the versions that work. If I don't pin this, we get an incompatible `@swc/core` package and the build errors.

File: tsconfig.node.json (hunk 2 above)
Comment (ryanclark) [reply tsconfig.node.json:10]: How'd you mean? This applies certain a certain tsconfig to these files, and is referenced by `tsconfig.json` in the root

File: tsconfig.json
Comment (ryanclark) [reply tsconfig.json:1]: I'll take a look

File: package.json (hunk 1 above)
Comment (ravicious) [reply package.json:28]: I see that master uses @vitejs/plugin-react-swc `^3.2.0` vs `^3.3.1` on this branch. I managed to massage yarn.lock into resolving @swc/core to 1.3.36 (a1bfb9f8c53e58) by doing this:

1. Remove @vitejs/plugin-react-swc from package json.
2. Run `yarn`.
3. Add it back to package.json with `"@vitejs/plugin-react-swc": "~3.2.0"` (this won't upgrade it past 3.2.x).
4. Run `yarn`.
5. Remove @swc/core resolution.

What do you think about about restricting plugin-react-swc to `~3.2.0` instead of adding that @swc/core resolution? I care about this only because adding a resolution is a pretty heavy-handed hammer that I think we should use only as a last resort.

**Comment by zmb3** [comment]
@ryanclark I'm going to close this one - it hasn't been touched in 4 months and eventually v12 will die out.

**Comment by ryanclark** [comment]
> @ryanclark I'm going to close this one - it hasn't been touched in 4 months and eventually v12 will die out.

Yeah fair enough. I didn't pick it back up as it seemed v12 wasn't going to be supported for much longer, I just forgot to close it
